src/charm.py

```diff
@@ -1,427 +1,127 @@
-        if interfaces["ingress"]:
-            interfaces["ingress"].send_data(
-                {
-                    "prefix": "/mlflow/",
-                    "rewrite": "/",
-                    "service": self.model.app.name,
-                    "port": self.model.config["mlflow_port"],
+    def _charmed_mlflow_layer(self) -> Layer:
+        """Create and return Pebble framework layer."""
+
+        layer_config = {
+            "summary": "mlflow-server layer",
+            "description": "Pebble config layer for mlflow-server",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entrypoint of mlflow-server image",
+                    "command": (
+                        "mlflow " "server " "--host " "0.0.0.0 " "--port " f"{self._port} "
+                    ),
+                    "startup": "enabled",
                 }
-            )
-
-    def _configure_mysql(
-        self,
-    ):
-        mysql = self.model.relations["db"]
-        if len(mysql) > 1:
-            raise CheckFailedError("Too many mysql relations", BlockedStatus)
+            },
+        }
 
-        try:
-            mysql = mysql[0]
-            unit = list(mysql.units)[0]
-            mysql = mysql.data[unit]
-            mysql["database"]
-            return mysql
-        except (IndexError, KeyError):
-            raise CheckFailedError("Waiting for mysql relation data", WaitingStatus)
+        return Layer(layer_config)
 
     def _check_leader(self):
+        """Check if this unit is a leader."""
         if not self.unit.is_leader():
-            # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailedError("Waiting for leadership", WaitingStatus)
-
-    def _get_interfaces(self):
+            self.logger.info("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _update_layer(self) -> None:
+        """Update the Pebble configuration layer (if changed)."""
+        current_layer = self.container.get_plan()
+        new_layer = self._charmed_mlflow_layer()
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info("Pebble plan updated with new configuration, replaning")
+                self.container.replan()
+            except ChangeError as err:
+                raise ErrorWithStatus(f"Failed to replan with error: {str(err)}", BlockedStatus)
+
+    def _on_pebble_ready(self, _):
+        """Configure started container."""
+        if not self.container.can_connect():
+            # Pebble Ready event should indicate that container is available
+            raise ErrorWithStatus("Pebble is ready and container is not ready", BlockedStatus)
+
```

Comment:
Should this be an Error rather than just something Blocked?  Blocked suggests it is something that a user can intervene about (change a setting, etc), but I don't think that's the case here

Reply:
Blocked is consistent with rest of our rewrites .... do we want to change it in every sidecar charm? If so Lets keep this Blocked and create a task. What do you think?

Reply:
I agree, a `BlockedStatus` means the charm needs a human operator to unblock, a `WaitingStatus` could be more appropriate here. I think this should be fixed before we merge your changes from the dev branch into main, but this PR is already very complex and convoluted. I am okay if you send a follow up PR just for this.